mods/taskbar-autohide-better.wh.cpp

```diff
@@ -0,0 +1,105 @@
+// ==WindhawkMod==
+// @id              taskbar-autohide-better
+// @name            Better Taskbar Autohide
+// @description     Allow taskbar autohide when inactive window notified
+// @version         1.0
+// @author          Cirn09
+// @github          https://github.com/cirn09
+// @homepage        https://blog.cirn09.xyz/
+// @include         explorer.exe
+// ==/WindhawkMod==
+
+// ==WindhawkModReadme==
+/*
+# Better Taskbar Autohide
+Allow taskbar autohide when inactive window notified
+![](https://p.sda1.dev/13/f403002dd34d5e45de384eb8eb626922/demo.gif)
+*/
+// ==/WindhawkModReadme==
+// ==WindhawkModSettings==
+/*
+- delay: 1500
+*/
+// ==/WindhawkModSettings==
+
+#include <minwindef.h>
+#include <windhawk_api.h>
+#include <string>
+#include <string_view>
+
+#ifdef _WIN64
+std::wstring_view WndProc_symname(
+    L"?v_WndProc@CTaskBand@@MEAA_JPEAUHWND__@@I_K_J@Z");
+std::wstring_view PermitAutoHide_symname(
+    L"?PermitAutoHide@CTaskListWnd@@UEAAHXZ");
+#else
+std::wstring_view WndProc_symname(
+    L"?v_WndProc@CTaskBand@@MAEJPAUHWND__@@IIJ@Z");
+std::wstring_view PermitAutoHide_symname(
+    L"?PermitAutoHide@CTaskListWnd@@UAGHXZ");
+#endif
+
+using WndProc_t = LRESULT(__thiscall*)(void* pThis,
+                                       HWND hWnd,
+                                       UINT msg,
+                                       WPARAM wParam,
+                                       LPARAM lParam);
+using PermitAutoHide_t = BOOL(__thiscall*)(void* pThis);
+
+WndProc_t WndProc_o = nullptr;
+PermitAutoHide_t PermitAutoHide_o = nullptr;
+FILETIME lastNotifyTime = {0};
+int delayMile_g;
+
+__thiscall LRESULT WndProc_hook(void* pThis,
+                                HWND hWnd,
+                                UINT msg,
+                                WPARAM wParam,
+                                LPARAM lParam) {
+    if (msg == 0xc028 && wParam == 0x8006 && lParam == 0x31270) {
+        GetSystemTimeAsFileTime(&lastNotifyTime);
+    }
+    return WndProc_o(pThis, hWnd, msg, wParam, lParam);
+}
+
+__thiscall BOOL PermitAutoHide_hook(void* pThis) {
+    FILETIME currentTime;
+    GetSystemTimeAsFileTime(&currentTime);
+    if (currentTime.dwLowDateTime - lastNotifyTime.dwLowDateTime > delayMile_g)
+        return TRUE;
+    else
+        return FALSE;
+}
+
+BOOL Wh_ModInit() {
+    Wh_Log(L"Init " WH_MOD_ID L" version " WH_MOD_VERSION);
+    WH_FIND_SYMBOL f;
+    HMODULE base = LoadLibrary(L"explorer.exe");
+    HANDLE hSymbol = Wh_FindFirstSymbol(base, NULL, &f);
```

Comment:
I suggest using `HookSymbols` of `windhawk_utils.h` for symbol resolution. It caches symbols to avoid loading and enumerating them every time, which can be slow. It's not documented properly yet but I plan on doing it soon, after releasing a new Windhawk version.

You'll have to use the undecorated version of the symbols, since that's what the API supports at this point, but I assume it's not a problem?

Usage example:
https://github.com/ramensoftware/windhawk-mods/blob/59d4fcd662a854e70f166ea1756cf92ca940b02e/mods/msg-box-font-fix.wh.cpp#L77-L91

Reply:
Actually, I used the decorated symbol because I not sure what the undecorated symbol was on x32 😂. The symbol given by windhawk is inconsistent with the result of IDA.
If possible, could you help me determine their undecorated symbol? 
```
#ifdef _WIN64
std::wstring_view WndProc_symname(
    L"?v_WndProc@CTaskBand@@MEAA_JPEAUHWND__@@I_K_J@Z");
std::wstring_view PermitAutoHide_symname(
    L"?PermitAutoHide@CTaskListWnd@@UEAAHXZ");
#else
std::wstring_view WndProc_symname(
    L"?v_WndProc@CTaskBand@@MAEJPAUHWND__@@IIJ@Z");
std::wstring_view PermitAutoHide_symname(
    L"?PermitAutoHide@CTaskListWnd@@UAGHXZ");
#endif
```

Reply:
Here you go. I wanted to have some tool to help with this for a while now. For now, I did something quick:

x64:
```
[00000000001F67E0] protected: virtual __int64 __cdecl CTaskBand::v_WndProc(struct HWND__ *,unsigned int,unsigned __int64,__int64)
[00000000001F67E0] ?v_WndProc@CTaskBand@@MEAA_JPEAUHWND__@@I_K_J@Z

[0000000000086EA0] public: virtual int __cdecl CTaskListWnd::PermitAutoHide(void)
[0000000000086EA0] ?PermitAutoHide@CTaskListWnd@@UEAAHXZ
```

x86:
```
[000000000009F5C0] protected: virtual long __thiscall CTaskBand::v_WndProc(struct HWND__ *,unsigned int,unsigned int,long)
[000000000009F5C0] ?v_WndProc@CTaskBand@@MAEJPAUHWND__@@IIJ@Z

[0000000000273BD0] public: virtual int __stdcall CTaskListWnd::PermitAutoHide(void)
[0000000000273BD0] ?PermitAutoHide@CTaskListWnd@@UAGHXZ
```

Reply:
Here, I finally made a tool for that:
https://github.com/m417z/windhawk-symbol-helper

Reply:
> Here, I finally made a tool for that: https://github.com/m417z/windhawk-symbol-helper

useful, but adding a browe path button will be better

Reply:
Maybe in a future update. For now I pushed a small update with automatic detection of Windhawk paths, bug fixes, and drag and drop support.